File: python-prototype/blockchain/blockchain/blockchain.py
from typing import Dict, List, Set

from .account_model import AccountModel
from .block import Block
from .transaction import Transaction, TxType
from .utils import Utils


class Blockchain():

    def __init__(self) -> None:
        self.blocks = [Block.genesis()]
        self.account_model = AccountModel()

    def to_json(self) -> Dict:
        json_dict = {}
        json_dict['blocks'] = [block.to_json() for block in self.blocks]
        return json_dict

    def latest_previous_hash(self) -> str:
        return Utils.hash(self.blocks[-1].payload()).hexdigest()

    def latest_block_height(self) -> int:
        return self.blocks[-1].height

    def is_valid_block_count(self, block: Block) -> bool:
        return self.blocks[-1].height == block.height - 1

    def is_valid_previous_block_hash(self, block: Block) -> bool:
        return self.latest_previous_hash() == block.previous_hash

    def add_block(self, block: Block) -> None:
        if self.is_valid_block_count(block) and self.is_valid_previous_block_hash(block):
            self.execute_transactions(block.transactions)
            self.blocks.append(block)

    def is_transaction_covered(self, transaction: Transaction) -> bool:
        # TODO remove echange transaction type check if type removed
        if transaction.tx_type == TxType.EXCHANGE:
            return True
        sender_balance = self.account_model.get_balance(
            transaction.sender_public_key)
        return sender_balance >= transaction.amount

    def get_covered_transactions(self, transactions: List[Transaction]) -> List[Transaction]:
        return [tx for tx in transactions if self.is_transaction_covered(tx)]

    def execute_transaction(self, transaction: Transaction) -> None:
        sender = transaction.sender_public_key
        receiver = transaction.receiver_public_key
        amount = transaction.amount
        self.account_model.update_balance(sender, -amount)
        self.account_model.update_balance(receiver, amount)

    def execute_transactions(self, transactions: List[Transaction]) -> None:
        for tx in transactions:
            self.execute_transaction(tx)
    
    def get_account_balance(self, public_key_string: str) -> int:
        return self.account_model.get_balance(public_key_string)

File: crypto-backend/blockchain/blockchain.go
package blockchain

import (
	"cryptomunt/proofOfStake"
	"cryptomunt/utils"
	"encoding/json"
	"errors"
)

type Blockchain struct {
	Blocks       []Block                    `json:"blocks"`
	AccountModel *AccountModel              `json:"-"`
	ProofOfStake *proofOfStake.ProofOfStake `json:"-"`
}

func CreateBlockchain() Blockchain {
	genesisBlock := CreateGenesisBlock()
	var blocks []Block
	blocks = append(blocks, genesisBlock)
	pos := proofOfStake.NewProofOfStake()

	accountModel := CreateAccountModel()

	return Blockchain{
		Blocks:       blocks,
		AccountModel: &accountModel,
		ProofOfStake: &pos,
	}
}

func (blockchain *Blockchain) ToJson() string {
	blockchainJson, err := json.Marshal(blockchain)
	if err != nil {
		panic("ERROR")
	}
	return string(blockchainJson)
}

func (blockchain *Blockchain) LatestBlockHeight() int {
	return len(blockchain.Blocks) - 1
}

func (blockchain *Blockchain) AddBlock(block Block) {
	if blockchain.IsValidBlockHeight(block) && blockchain.IsValidPreviousBlockHash(block) {
		blockchain.ExecuteTransactions(block.Transactions)
		blockchain.Blocks = append(blockchain.Blocks, block)
	}
}

func (blockchain *Blockchain) ExecuteTransactions(transactions []Transaction) {
	for _, transaction := range transactions {
		blockchain.executeTransaction(transaction)
	}
}

func (blockchain *Blockchain) executeTransaction(transaction Transaction) {
	sender := transaction.SenderPublicKey
	receiver := transaction.ReceiverPublicKey
	amount := transaction.Amount
	amountWithoutFee := CalculateInitialAmount(amount)

	if transaction.Type == STAKE {
		if sender == receiver {
			blockchain.ProofOfStake.UpdateStake(sender, amountWithoutFee)
			blockchain.AccountModel.UpdateBalance(sender, -amount)
		}
	} else if transaction.Type == REWARD {
		blockchain.AccountModel.UpdateBalance(receiver, amount)
	} else {
		blockchain.AccountModel.UpdateBalance(sender, -amount)
		blockchain.AccountModel.UpdateBalance(receiver, amountWithoutFee)
	}
}

func (blockchain *Blockchain) IsValidBlockHeight(block Block) bool {
	blockLength := len(blockchain.Blocks)
	return blockchain.Blocks[blockLength-1].Height == block.Height-1
}

func (blockchain *Blockchain) IsValidPreviousBlockHash(block Block) bool {
	return blockchain.LatestPreviousHash() == block.PreviousHash
}

func (blockchain *Blockchain) IsValidForger(block Block) bool {
	return block.Forger == blockchain.getNextForger()
}

func (blockchain *Blockchain) IsBlockTransactionsValid(block Block) bool {
	transactions := block.Transactions
	coveredTransactions := blockchain.GetCoveredTransactions(transactions)
	return len(transactions) == len(coveredTransactions)
}

func (blockchain *Blockchain) IsBlockRewardTransactionValid(block Block) bool {
	if !blockHasOneRewardTransaction(block) {
		return false
	}

	rewardTx, _ := getRewardTransactionFromBlock(block)
	return rewardTransactionHasCorrectReceiver(block, rewardTx) && rewardTransactionHasCorrectAmount(block, rewardTx)
}

func blockHasOneRewardTransaction(block Block) bool {
	rewardTxFound := false

	for _, transaction := range block.Transactions {
		if transaction.Type == REWARD && rewardTxFound {
			return false
		}
		if transaction.Type == REWARD {
			rewardTxFound = true
		}
	}

	return rewardTxFound
}

func getRewardTransactionFromBlock(block Block) (Transaction, error) {
	for _, transaction := range block.Transactions {
		if transaction.Type == REWARD {
			return transaction, nil
		}
	}
	return *new(Transaction), errors.New("No reward transaction in block")
}

func rewardTransactionHasCorrectReceiver(block Block, rewardTx Transaction) bool {
	return block.Forger == rewardTx.ReceiverPublicKey
}

func rewardTransactionHasCorrectAmount(block Block, rewardTx Transaction) bool {
	transactionsLength := len(block.Transactions)
	totalReward := CalculateTotalReward(block.Transactions[:transactionsLength-1])
	return rewardTx.Amount == totalReward
}

func (blockchain *Blockchain) LatestPreviousHash() string {
	blockLenght := len(blockchain.Blocks)
	latestBlock := blockchain.Blocks[blockLenght-1]
	payload := latestBlock.Payload()

	return utils.GetHexadecimalHash(payload)
}

func (blockchain *Blockchain) GetCoveredTransactions(transactions []Transaction) []Transaction {
	var coveredTransactions = make([]Transaction, len(transactions))

	for _, transaction := range transactions {
		if blockchain.IsTransactionCovered(transaction) {
			coveredTransactions = append(coveredTransactions, transaction)
		}
	}
	return coveredTransactions
}

func (blockchain *Blockchain) IsTransactionCovered(transaction Transaction) bool {
	if transaction.Type == EXCHANGE || transaction.Type == REWARD {
		return true
	}
	senderBalance := blockchain.AccountModel.GetBalance(transaction.SenderPublicKey)
	return senderBalance >= transaction.Amount
}

func (blockchain *Blockchain) GetAccountBalance(publicKey string) int {
	return blockchain.AccountModel.GetBalance(publicKey)
}

func (blockchain *Blockchain) getNextForger() string {
	prevBlockHash := blockchain.LatestPreviousHash()
	return blockchain.ProofOfStake.PickForger(prevBlockHash)
}
